### PR TITLE
more and improved error tracing

### PR DESCRIPTION
**Pull request by pfitzseb**
Fixes https://github.com/julia-vscode/julia-vscode/pull/1423#issuecomment-651915275.

### CODE REVIEW

File: src/interactive/modules.ts [hunk 1]
```diff
@@ -91,16 +93,24 @@ async function updateModuleForEditor(editor: vscode.TextEditor) {
...
         mod = await getModuleForEditor(editor)
     } catch (err) {
-        console.error(err)
-        return
+        if (g_languageClient) {
+            telemetry.traceTrace({
+                message: err.toString(),
+                severity: SeverityLevel.Error
+            })
+        }
```
Comment (davidanthoff) [inline src/interactive/modules.ts:101]: The idea is that this is a proper bug, right? I think we should use `handleNewCrashReport` to send an exception in that case, not a trace. We might have to add another version of it that accepts an exception object directly.

There are three benefits: 1) it shows up in failures, where it belongs :), 2) it includes stack traces, 3) it honors the opt-in for crash reporting.

Oh, and I think we should actually implement something like https://github.com/julia-vscode/julia-vscode/pull/1276, and then let the error bubble up, i.e. we should add generic crash reporting to the extension code, and then let real errors bubble up to them. So in this case, we should probably just rethrow the exception if `g_languageClient` is not `undefined`.

The reason I'm using traces for the these position validation things is that right now I'm still not sure what exactly is going wrong, i.e. whether we are making an error or not etc.

File: src/interactive/modules.ts [hunk 2]
```diff
@@ -91,16 +93,24 @@ async function updateModuleForEditor(editor: vscode.TextEditor) {
...
         loaded = await g_connection.sendRequest(requestTypeIsModuleLoaded, mod)
     } catch (err) {
-        console.error(err)
-        return
+        if (g_connection) {
+            telemetry.traceTrace({
+                message: err.toString(),
+                severity: SeverityLevel.Error
+            })
+        }
```
Comment (davidanthoff) [inline src/interactive/modules.ts:113]: Same here, I think this should show up as an exception, not a trace, and we should add a higher level error handling like in the other PR to catch such errors.

File: src/interactive/modules.ts [hunk 3]
```diff
@@ -111,8 +121,14 @@ async function chooseModule() {
     try {
         possibleModules = await g_connection.sendRequest(requestTypeGetModules, null)
     } catch (err) {
-        console.error(err)
-        vscode.window.showInformationMessage('Setting a module requires an active REPL.')
+        if (g_connection) {
+            telemetry.traceTrace({
+                message: err.toString(),
+                severity: SeverityLevel.Error
+            })
```
Comment (davidanthoff) [inline src/interactive/modules.ts:128]: Same :)

File: src/interactive/modules.ts (hunk 1 above)
Comment (pfitzseb) [reply src/interactive/modules.ts:97]: Done. I don't think we can really get stacktraces here though. Also not sure what the latest argument is supposed to be :)

Comment (davidanthoff) [reply src/interactive/modules.ts:97]: I I pushed two more commits to this PR: one adds a crash reporting function that accepts a JavaScript `Error` object, and then the stack trace gets automatically extracted, the second one uses this here.

The last argument is the role, it allows us to filter on errors, essentially we use it to signal in which process the crash happened. Right now we use the values `Debugger`, `Extension`, `Language Server`, `REPL` and `Symbol Server`.

Comment (pfitzseb) [reply src/interactive/modules.ts:97]: Ideally we'd use the _Julia stacktrace_ here, but we can figure out a good way to do that later.

**Pull request merged**